### PR TITLE
Module 4

### PR DESCRIPTION
**Pull request by MaickMos**
# Push Notificactions  module #4
En esta PR se hara entrega de la continuacion del modelo predictivo para posibles personas interesadas en un producto. Dado un usuario y producto, se predice si compraria ese producto si estuvieran comprando en ese momento. Condiciones: los usuarios deben ser aquellos que realizan una compra de minimo 5 productos. Esta vez usaron modelos no lineales 

## Preparacion del dataset
Se usaron el mismos dataset de la practica anterior y una variacion donde se obesrvaron las principales caracteristicas que toma un random forest para escoger las mas importantes, quedando con 9
## Metricas de evaluacion
Se aplicaron las misma metricas de evaluacion dentro de una funcion y se implemento una nueva con las metricas de 
- roc_auc_score
- log_loss
- average_precision_score
Se conservan las anteriores:
- Recall
- precision-recall curve
- ROC Curve

## Modelos
Se concervan los mejores modeles de baseline y regresion logistica para futuras compraciones
### Random Forest
Se realiza una prueba con el random forest con el dataset completo, se obtienen resultados malos con una average_precision de 0.014 y unAUC en precision recall curve de 0.02 para el dataset de test.
Se escogen las feautures con un valor mayor a 3%, se crea un nuevo data set con 9 variables.
Se compara este dataset con el anterior de 3 varaibles.

image

Se obtienen mejores resultados, siendo el mejor el de tres variables en test con un AUC de 0.10 y  average_precision de 0.027.
por ultimo se testea con el numero de arboles de 5, 10 y 100, usando el dataset de 3 variables.
Se obtievieron resultados similares en test siendo el mejor el de 10 arboles, con AUC de 0.14 y average precision de 0.046.
Se seleciona este ultimo modelo ya que tiene un rendimiento similar al de 100 de arboles que tiene el modelo por defecto.

### Gradient Boosting Trees
Se realiza una prueba con hiperparametro de n_estimators=100, learning_rate=0.05, y max_depth=5.
Se obtiene resultados similares al de random forest.
Se realiza una variacion entre el learning rate y la profundida de los arboles.

image

El que mejor rendimiento tiene es de 0.05 learn rate y 5 de profundidad.
### Comparacion
Se comparan los 4 cuatro modelos y se concluye.

image

La regresión logística con regularización Ridge fue el mejor modelo en este caso para valores de recall en el rango de 0,1 a 0,3, logrando un AUC de 0,16 en la curva de precisión-recall.

Se probaron diferentes versiones del conjunto de datos: el conjunto de datos numérico completo, el conjunto de datos con solo coeficientes superiores a 0,03 y el conjunto de datos con las tres variables más importantes. En todos los modelos, el mejor rendimiento se obtuvo utilizando las tres variables más importantes. Irónicamente, esto también fue cierto para los modelos basados ​​en árboles, pero ningún ajuste mejoró su rendimiento significativamente.

Todos los modelos superaron la línea base, lo que confirma su valor predictivo. La regresión logística sigue siendo la mejor opción en este caso. Quizás los modelos basados ​​en árboles requieren un ajuste más profundo de los hiperparámetros, o la relación entre las variables es más lineal, lo que favorece la regresión.

# Archivo para cargar el modelo
## Fit
Se creo el script de fit.py, permite cargar la parametrizacion para un modelo de regresion logistica con la fecha y hora generada y lo guarda en un path, retorna el lugar(path) donde se guardo
## Predict
script que llama a un modelo guardado en un path, transforma los datos para ingresarlos en el modelo y obtener las prediciones.

Code Writer: @sergiorozada12

### CODE REVIEW

File: src/module_4/filtered_users.json
Comment (sergiorozada12) [inline src/module_4/filtered_users.json:1]: It is recommended to put config files / data into separate folders to be clear about the content of them

File: src/module_4/fit.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+def Standardization(x_train: pd.DataFrame):
+    scaler = StandardScaler()
+    x_train_columns = x_train.columns
+    x_train = scaler.fit_transform(x_train)
+    x_train = pd.DataFrame(x_train, columns=x_train_columns)
+    return x_train
+
+
+def slip_data_set(df, y):
```
Comment (sergiorozada12) [inline src/module_4/fit.py:69]: Pay attention to the typings. In some functions you type, in others you don't. Be consistent.

File: src/module_4/fit.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    model_path = save_model_with_date(model, event)
+
+    return {"statusCode": 200, "body": json.dumps({"model_path": model_path})}
+
+
+def main():
+    event = {
+        "model_parametrisation": {"penalty": "l2", "solver": "liblinear"},
+        "dataset_path": "D:/Users/maick/Desktop/Codigos/zrive-ds/data/box_builder_dataset/feature_frame.csv",
```
Comment (sergiorozada12) [inline src/module_4/fit.py:127]: Move all this fixed paths to the beginning of the script with capitalized names. If this would be a PR in a working environment, I would be rejecting the PR. Also, use always relative paths.

File: src/module_4/fit.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            raise ValueError("Missing required columns in dataset!")
+        return df
+
+    except FileNotFoundError as e:
+        print(f"Error not found: {file_path}")
+
+    except ValueError as e:
+        print(e)
+    except Exception as e:
+        print(f"Error loading data: {e}")
```
Comment (sergiorozada12) [inline src/module_4/fit.py:57]: Beware of prints. Use always logging in python. It's the same effort and it scales way better to pro environments.

File: src/module_4/fit.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+def load_data(file_path: str):
+    try:
+
+        if not os.path.exists(file_path):
+            raise FileNotFoundError(f"File {file_path} does not exist.")
+
+        df = pd.read_csv(file_path)
+
+        REQUIRED_COLUMNS = {
```
Comment (sergiorozada12) [inline src/module_4/fit.py:25]: This fix variables are never declared inside functions. Move it to the beginning of the file.

File: src/module_4/predict.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+from pathlib import Path
+
+
+def handler_predict(event, _):
+    try:
+        model_files = sorted(os.listdir(event["model_path"]), reverse=True)
+
+        if not model_files:
+            raise FileNotFoundError("No models trained.")
+        # Select the lastest
```
Comment (sergiorozada12) [inline src/module_4/predict.py:14]: This comments can go out. Why would you want to keep them here?

File: src/module_4/predict.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import os
+import pickle
+import pandas as pd
+import json
+from pathlib import Path
+
+
+def handler_predict(event, _):
```
Comment (sergiorozada12) [inline src/module_4/predict.py:8]: Why this second empty argument? Get rid of it if you do not need it

File: src/module_4/predict.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            result[user_id] = float(pred)
+
+        return {"statusCode": 200, "body": json.dumps({"prediction": result})}
+
+    except Exception as e:
+        return {"statusCode": 500, "body": json.dumps({"error": str(e)})}
+
+
+def main():
+    JSON_PATH = Path(
```
Comment (sergiorozada12) [inline src/module_4/predict.py:37]: Same, these variables always to the beginning of the file. Use relative paths.

File: src/module_4/take_100_users.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+
+    OUTPUT_PATH = Path("D:\Users\maick\Desktop\Codigos\zrive-ds\src\module_4\models")
+
+
+    output_path = OUTPUT_PATH / "filtered_users.json"
+    with open(output_path, "w", encoding="utf-8") as f:
+        json.dump(result, f, indent=2)
+
+    print(f"Archivo guardado en: {output_path}")
```
Comment (sergiorozada12) [inline src/module_4/take_100_users.py:22]: Do not use Spanish when coding. You neer know when your code is going to be read by a non-Spanish-speaking person.

**Review by sergiorozada12 — APPROVED** [review]
Nice PR. I really liked the modeling NB. You were very exhaustive at the evaluation. I left you some comments in the code. You should put more attention to the small things that make your code more solid and proffessional. Merge whenever you want, I leave you here a couple of things about the NB:
* The performance of the models is not so good overall. Not your fault, the data is not the best. Do you think you have stratifications of the data where the model can perform better? Maybe in a pro environment we only need to model top clients and leave some of the data appart.
*  How would you use this model for a business case? What would you optimize here? What business goal we can try to achieve with this model?
